### PR TITLE
Add Python 3.11 Support

### PR DESCRIPTION
**Pull request by fa9r**
## Describe changes
Add (limited) support for Python 3.11.

In the process, we also relaxed the version requirements for `tensorflow`, `tensorboard`, `transformers`, and `torchvision`.

### Limited Support
The following integrations are currently not supported under 3.11:
- gcp
- kubeflow
- tekton

This is because because:
- GCP packages that support Python 3.11 are not compatible with KFP 1
- We cannot upgrade to KFP 2 (#1697) because Tekton doesn't have any release compatible with KFP 2 yet

Previous PR: #1701 

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Comment by strickvl** [comment]
@fa9r is the description for this PR still valid / up-to-date? if not, could you update with current status?

**Comment by fa9r** [comment]
@strickvl updated PR description 👍

File: tests/integration/examples/test_tensorflow.py [hunk 1]
```diff
@@ -12,12 +12,17 @@
...
+import sys
 
 import pytest
 
 from tests.integration.examples.utils import run_example
 
 
+@pytest.mark.skipif(
+    sys.version_info > (3, 10),
+    reason="Kubeflow integration not installed in python 3.11",
```
Comment (strickvl) [inline tests/integration/examples/test_tensorflow.py:24]:
```suggestion
    reason="Kubeflow integration not installed in Python 3.11",
```

File: tests/unit/integrations/gcp/artifact_stores/test_gcp_artifact_store.py [hunk 2]
```diff
@@ -42,6 +44,10 @@ def _get_gcp_artifact_store(**kwargs):
     )
 
 
+@pytest.mark.skipif(
+    sys.version_info > (3, 10),
+    reason="GCP integration not installed in python 3.11",
```
Comment (strickvl) [inline tests/unit/integrations/gcp/artifact_stores/test_gcp_artifact_store.py:49]:
```suggestion
    reason="GCP integration not installed in Python 3.11",
```

File: tests/unit/integrations/gcp/image_builders/test_gcp_image_builder.py [hunk 3]
```diff
@@ -42,6 +48,10 @@ def _get_gcp_image_builder(
     )
 
 
+@pytest.mark.skipif(
+    sys.version_info > (3, 10),
+    reason="GCP integration not installed in python 3.11",
```
Comment (strickvl) [inline tests/unit/integrations/gcp/image_builders/test_gcp_image_builder.py:53]:
```suggestion
    reason="GCP integration not installed in Python 3.11",
```

File: tests/unit/integrations/gcp/orchestrators/test_vertex_orchestrator.py [hunk 4]
```diff
@@ -49,13 +49,22 @@ def _get_vertex_orchestrator(**kwargs):
     )
 
 
+@pytest.mark.skipif(
+    sys.version_info > (3, 10),
+    reason="GCP integration not installed in python 3.11",
```
Comment (strickvl) [inline tests/unit/integrations/gcp/orchestrators/test_vertex_orchestrator.py:54]:
```suggestion
    reason="GCP integration not installed in Python 3.11",
```

File: tests/unit/integrations/gcp/orchestrators/test_vertex_orchestrator.py [hunk 5]
```diff
@@ -129,6 +138,10 @@ def test_vertex_orchestrator_stack_validation(
         ).validate()
 
 
+@pytest.mark.skipif(
+    sys.version_info > (3, 10),
+    reason="GCP integration not installed in python 3.11",
```
Comment (strickvl) [inline tests/unit/integrations/gcp/orchestrators/test_vertex_orchestrator.py:143]:
```suggestion
    reason="GCP integration not installed in Python 3.11",
```

File: tests/unit/integrations/kubeflow/orchestrators/test_kubeflow_orchestrator.py [hunk 6]
```diff
@@ -50,6 +56,10 @@ def _get_kubeflow_orchestrator(
     )
 
 
+@pytest.mark.skipif(
+    sys.version_info > (3, 10),
+    reason="Kubeflow integration not installed in python 3.11",
```
Comment (strickvl) [inline tests/unit/integrations/kubeflow/orchestrators/test_kubeflow_orchestrator.py:61]:
```suggestion
    reason="Kubeflow integration not installed in Python 3.11",
```

File: tests/unit/integrations/kubeflow/orchestrators/test_kubeflow_orchestrator.py [hunk 7]
```diff
@@ -93,6 +103,10 @@ def test_kubeflow_orchestrator_remote_stack(
         ).validate()
 
 
+@pytest.mark.skipif(
+    sys.version_info > (3, 10),
+    reason="Kubeflow integration not installed in python 3.11",
```
Comment (strickvl) [inline tests/unit/integrations/kubeflow/orchestrators/test_kubeflow_orchestrator.py:108]:
```suggestion
    reason="Kubeflow integration not installed in Python 3.11",
```

File: tests/unit/integrations/kubeflow/test_utils.py [hunk 8]
```diff
@@ -13,14 +13,22 @@
...
+
+import pytest
 
 from zenml.integrations.kubeflow.utils import apply_pod_settings
 from zenml.integrations.kubernetes.pod_settings import KubernetesPodSettings
 
 
+@pytest.mark.skipif(
+    sys.version_info > (3, 10),
+    reason="Kubeflow integration not installed in python 3.11",
```
Comment (strickvl) [inline tests/unit/integrations/kubeflow/test_utils.py:26]:
```suggestion
    reason="Kubeflow integration not installed in Python 3.11",
```

File: tests/unit/integrations/mlflow/experiment_trackers/test_mlflow_experiment_tracker.py [hunk 9]
```diff
@@ -107,9 +109,9 @@ def test_mlflow_experiment_tracker_stack_validation() -> None:
         created=datetime.now(),
         updated=datetime.now(),
     )
-    remote_artifact_store = GCPArtifactStore(
+    remote_artifact_store = S3ArtifactStore(
         name="",
-        config=GCPArtifactStoreConfig(path="gs://my-bucket"),
+        config=S3ArtifactStoreConfig(path="s3://tmp"),
```
Comment (strickvl) [inline tests/unit/integrations/mlflow/experiment_trackers/test_mlflow_experiment_tracker.py:114]: how is the flavor (below) gcp if the config has been changed to s3?

Comment (fa9r) [reply tests/unit/integrations/mlflow/experiment_trackers/test_mlflow_experiment_tracker.py:114]: Nice catch, I changed flavor to "s3" accordingly. Didn't seem to have any effect actually.

**Review by strickvl — APPROVED** [review]

**Comment by stefannica** [comment]
We're running into https://github.com/pytorch/pytorch/issues/99637 after removing the tensorflow version pin. Trying with tensorflow pinned down to <=2.11 again.

**Review by stefannica — APPROVED** [review]

**Pull request merged**